### ISSUE TITLE
WebCrypto roadmap

### ISSUE THREAD

**Issue by lucacasonato** [denoland/deno#11690]
Spec: https://w3c.github.io/webcrypto

This is the global tracking issue for WebCrypto API implementation.

See the status of the entire API surface:

- [x] Enable non-extractable keys: #11481
- [x] `SubtleCrypto.encrypt`
  - [x] `RSA-OAEP`
  - [x] `AES-CTR` https://github.com/denoland/deno/pull/13177
  - [x] `AES-CBC`
  - [x] `AES-GCM`
- [x] `SubtleCrypto.decrypt`
  - [x] `RSA-OAEP`
  - [x] `AES-CTR` https://github.com/denoland/deno/pull/13177
  - [x] `AES-CBC`
  - [x] `AES-GCM`
- [x] `SubtleCrypto.sign`
  - [x] `RSASSA-PKCS1-v1_5`
  - [x] `RSA-PSS` (all hashes)
  - [x] `RSA-PSS` (all hashes)
  - [x] `ECDSA`
    - [x] `P256`
    - [x] `P384`
    - [ ] `P521`
  - [x] `HMAC`
- [x] `SubtleCrypto.verify`
  - [x] `RSASSA-PKCS1-v1_5`
  - [x] `RSA-PSS` (all hashes)
  - [x] `RSA-PSS` (all hashes)
  - [x] `ECDSA`
    - [x] `P256`
    - [x] `P384`
    - [ ] `P521`
  - [x] `HMAC`
- [x] `SubtleCrypto.digest`
  - [x] `SHA-1`
  - [x] `SHA-256`
  - [x] `SHA-384`
  - [x] `SHA-512`
- [x] `SubtleCrypto.generateKey`
  - [x] `RSASSA-PKCS1-v1_5`
  - [x] `RSA-PSS`
  - [x] `RSA-PSS`
  - [x] `RSA-OAEP` 
  - [x] `ECDSA`
    - [x] `P256`
    - [x] `P384`
    - [ ] `P521`
  - [x] `ECDH` #11870
    - [x] `P256`
    - [x] `P384`
    - [ ] `P521`
  - [x] `AES-CTR` #11869
  - [x] `AES-CBC` #11869
  - [x] `AES-GCM` #11869
  - [x] `AES-KW` #11869
  - [x] `HMAC`
- [x] `SubtleCrypto.deriveKey`
  - [x] `ECDH`
  - [x] `HKDF`
  - [x] `PBKDF2`
- [x] `SubtleCrypto.deriveBits`
  - [x] `ECDH`
  - [x] `HKDF`
  - [x] `PBKDF2`
- [x] `SubtleCrypto.importKey`
  - [x] `RSASSA-PKCS1-v1_5`
    - [x] spki
    - [x] pkcs8
    - [x] jwk
  - [x] `RSA-PSS`
    - [x] spki
    - [x] pkcs8
    - [x] jwk
  - [x] `RSA-OAEP` 
    - [x] spki
    - [x] pkcs8
    - [x] jwk
  - [x] `ECDSA` (missing p512 curve)
    - [x] spki
    - [x] pkcs8
    - [x] jwk
    - [x] raw
  - [x] `ECDH` (missing p512 curve)
    - [x] spki
    - [x] pkcs8
    - [x] jwk
    - [x] raw
  - [x] `AES-CTR`
    - [x] jwk
    - [x] raw
  - [x] `AES-CBC`
    - [x] jwk
    - [x] raw
  - [x] `AES-GCM`
    - [x] jwk
    - [x] raw
  - [x] `AES-KW`
    - [x] jwk
    - [x] raw
  - [x] `HMAC`
    - [x] jwk
    - [x] raw
  - [x] `HKDF`
    - [x] raw
  - [x] `PBKDF2`
    - [x] raw
- [x] `SubtleCrypto.exportKey`
  - [x] `RSASSA-PKCS1-v1_5`
    - [x] spki
    - [x] pkcs8
    - [x] jwk 
  - [x] `RSA-PSS`
    - [x] spki
    - [x] pkcs8
    - [x] jwk 
  - [x] `RSA-OAEP` 
    - [x] spki
    - [x] pkcs8
    - [x] jwk 
  - [x] `ECDSA` (missing p512 curve)
    - [x] spki
    - [x] pkcs8
    - [x] jwk
    - [ ] raw
  - [x] `ECDH` (missing p512 curve)
    - [x] spki
    - [x] pkcs8
    - [x] jwk
    - [ ] raw
  - [x] `AES-CTR`
    - [x] raw
    - [x] jwk
  - [x] `AES-CBC`
    - [x] raw
    - [x] jwk
  - [x] `AES-GCM`
    - [x] raw
    - [x] jwk
  - [x] `AES-KW`
    - [x] raw
    - [x] jwk
  - [x] `HMAC`
- [x] `SubtleCrypto.wrapKey`
  - [x] `RSA-OAEP`
  - [x] `AES-CTR`
  - [x] `AES-CBC`
  - [ ] `AES-GCM`
  - [x] `AES-KW`
- [ ] `SubtleCrypto.unwrapKey`
  - [x] `RSA-OAEP`
  - [x] `AES-CTR`
  - [x] `AES-CBC`
  - [ ] `AES-GCM`
  - [x] `AES-KW`

**Comment by johnspurlock** [comment]
Please consider adding ED25519 to this list.

Cloudflare Workers [recently added](https://community.cloudflare.com/t/2021-6-25-workers-runtime-release-notes/281775) this [non-yet-standard](https://www.chromestatus.com/feature/4913922408710144) algorithm since it was requested so often by developers writing [Discord integrations](https://discord.com/developers/docs/interactions/receiving-and-responding#security-and-authorization).

The [operations they support](https://developers.cloudflare.com/workers/runtime-apis/web-crypto#supported-algorithms) are `sign`, `verify`, `generateKey`, `exportKey`, and `importKey`.

Support for this would be nice for folks wanting to write cross-worker-platform code with similar crypto support and performance characteristics.

**Comment by lucacasonato** [comment]
I will bring this up with the webcrypto working group at W3C. It seems like a logical addition.

**Comment by panva** [comment]
I applaud this overview. Thank you

**Comment by panva** [comment]
As far as urgency goes, given my wishful target of supporting JWS formatted JWTs first the most important missing piece is SubtleCrypto.importKey jwk, spki, and pkcs8 format support for RSASSA-PKCS1-v1_5, RSA-PSS, and ECDSA keys.

**Comment by panva** [comment]
@littledivy I'm working on a deno-specific publish and build process for [`jose`](https://github.com/panva/jose) - that is pure TS export and autopublish to https://deno.land/x/jose.

While the implementation is incomplete it is not possible to use the package without `--no-check` due to the missing types.

Would it be possible to make it so that the WebCrypto **interface** would be complete and type-defined whilst the actual missing implementations simply throwing `throw new DOMException("Not implemented", "NotSupportedError")`?

It's global type interfaces such as `HmacKeyAlgorithm`, `RsaKeyAlgorithm` completely missing as well as `importKey` being hardcoded to format `"raw"` at the moment rather than the `KeyFormat` enum.

**Comment by littledivy** [comment]
@panva Yeah missing types can be annoying but I think that they should match reality. cc @lucacasonato 

btw `HmacKeyAlgorithm`, `RsaKeyAlgorithm` and `EcKeyAlgorithm` should be fixed in #11738

also `lib.dom.d.ts` has harcoded "raw" in `importKey` while `KeyFormat` is declared but not used in the decl https://github.com/denoland/deno/blob/91051047cc82e6074348d17414854efb906019d2/cli/dts/lib.dom.d.ts#L15258

**Comment by panva** [comment]
I would very much prefer the types to be complete to allow for portable implementations that went out of their way to be fully built using TS and prepared to be consumed in Deno, rather than having to instruct users to disable their type precious type checking.

**Comment by panva** [comment]
ad importKey, that would appear to be a mistake on `lib.dom.d.ts` part, see https://www.w3.org/TR/WebCryptoAPI/#subtlecrypto-interface

**Comment by yacinehmito** [comment]
> @lucacasonato the table is lacking `SubtleCrypto.importKey` breakdown by `format` value. It would appear importKey is almost done when that's not the case.

@lucacasonato Seconded. `importKey` is barely done.

**Comment by 89z** [comment]
https://github.com/denoland/deno/issues/1891

**Comment by oles** [comment]
> Thank you for making steady progress here @littledivy.
> 
> I'd like to relay a wish of some of my modules' users. They would love it if JWK/SPKI/PKCS8 (in that order) import work was prioritized for RSA-PSS, ECDSA, and RSASSA-PKCS1-v1_5 so that at least signed JWTs functionality would be unlocked for them.

Seconded, I've been trying to use the Apple App Store Connect API with Deno, but I'm stuck on signing the JWT, and I think this is my blocker.

I'd love to not having to fall back to Node.js on this one, as I'm very in love with Deno Deploy!


I'd also love any suggestions on how to solve this meanwhile, as I've had no luck with the hits on https://deno.land/x?query=jwt, and neither highly experimental attempts with https://deno.land/x/god_crypto and https://deno.land/x/secp256k1

My plan for tomorrow is to check out https://github.com/PeculiarVentures/webcrypto and see if it can work with Deno, before I have to go some other route.

<br>

Here's a gist of a Node.js version that I found highly useful if anyone else stumbles upon this in the future:
https://gist.github.com/FelDev/50d969fe7249da949ed966538929310b

It uses "jsonwebtoken", which uses "jws", which uses "jwa", which finally use the built-in node crypto module createSign method, that is missing(?) in https://deno.land/std/node/crypto.ts for now

**Comment by panva** [comment]
@oles once #13013 lands and gets released you'll be able to use https://deno.land/x/jose with the key material you have. Until then there's just no way.

**Comment by oles** [comment]
@panva thank you for the notice! 

@SeanWykes let me know if you need a tester, as there's likely not much else I can help with when it comes to crypto, but I have at least have become a sponsor of the project :+1:

**Comment by SeanWykes** [comment]
> @SeanWykes let me know if you need a tester, as there's likely not much else I can help with when it comes to crypto, but I have at least have become a sponsor of the project 👍

@oles thanks for the offer, it'll be great if you can test #13013 - the PR includes JWK for EC and RSA - you can find the source branch [here](https://github.com/SeanWykes/deno/tree/crypto-ec-import-export) if you're ok with building deno, otherwise, let me know.

I've done some preliminary testing of RSA-PSS JWK using [jose](https://deno.land/x/jose@v4.3.7).

**Comment by panva** [comment]
@SeanWykes @oles needs P-256 ECDSA to sign an ES256 JWT.

**Comment by SeanWykes** [comment]
> @SeanWykes @oles needs P-256 ECDSA to sign an ES256 JWT. I already know the sign/verify operations work fine in `jose`, its just the key material imports that are missing for JWS/JWT to kick in and be usable.

So should be ready to go, the JWK key imports are already implemented in #13013.

**Comment by panva** [comment]
@SeanWykes nice. In case you want to run the full suite of jose tests under Deno using the latest release, this is how you might do that.

```
git clone --depth 1 --branch v4.3.7 https://github.com/panva/jose.git
cd jose
deno test --reload --jobs --allow-net --allow-read test-deno # replace deno with your local build
```

At the moment, with deno 1.16.4, the results look like this: `test result: FAILED. 73 passed; 73 failed`

**Comment by oles** [comment]
@panva success over at https://github.com/denoland/deno/pull/13013 with both key import and signing!

Thank you for the pointers and the library, and thank you @SeanWykes for your work on this - let me know if there's something more you'd like for me to test or help out with :+1:

**Comment by lucacasonato** [comment]
@panva JOSE test suite on current `main` branch: `test result: FAILED. 100 passed; 46 failed`.

ECDSA & ECDH imports / exports are coming very soon.

**Comment by panva** [comment]
@lucacasonato I was checking that yesterday, great progress! much props to yourself, @SeanWykes, @littledivy and anyone else participating in the effort.

I'd like to think the jose test suite is a good interoperability resource in addition to the web platform tests.

**Comment by littledivy** [comment]
[~98.1%](https://wpt.fyi/results/WebCryptoAPI?label=master&label=experimental&product=chrome&product=deno&product=firefox&product=safari&aligned) of the WebCrypto APIs are now part of Deno v1.18.0 :tada:

Thank you to @SeanWykes and everyone involved in this nearly 6 month long effort! 
Let's open new issue trackers for pending features (like p512 curve and non 96 bit ivs). Closing :-)

**Comment by panva** [comment]
Let me join in the congratulations. @littledivy @SeanWykes @lucacasonato thank you for these amazing efforts!

---

From a pending features perspective (JOSE completeness-wise) I'm looking at general P-521 support, P-384 ECDH. It'd be great to plan for [Curve25519 and Curve448](https://github.com/twiss/webcrypto-secure-curves), either through adding this draft support or following Node.js and CF Workers in adding the NODE- prefixed versions until said draft is adopted as an official extension.

**Comment by SeanWykes** [comment]
And here's my personal thank you to @littledivy, @lucacasonato and @bartlomieju for your patience and technical guidance for a Deno / Rust newbie. I've learnt a lot from you guys!

**Comment by diachedelic** [comment]
It looks like the only things missing are P-521 support and support for wrapping/unwrapping AES-GCM keys. Is anybody working on P-521 support? I can give it a go.

**Comment by diachedelic** [comment]
@littledivy It appears that [the issue](https://github.com/RustCrypto/elliptic-curves/issues/240) mentioned in [your TODO comment](https://github.com/denoland/deno/blob/7eee521199e9735ab7c347d99e9d90ba3046be1a/ext/crypto/lib.rs#L561) has just been resolved, meaning that it should be possible to make SubtleCrypto.deriveBits work with P-384 keys.
